### PR TITLE
refactor interfaces

### PR DESCRIPTION
**Pull request by andiwand**
- extract legacy ms code into its own module
- refactor frontend (`Reader` is dead, long live `Document`)
- add frontend with exceptions
- refactor `Storage`

### CODE REVIEW

File: common/include/common/Document.h [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+class Document {
+public:
+  virtual ~Document() = default;
+
+  virtual const FileMeta &meta() const noexcept = 0;
+
+  virtual bool decrypted() const noexcept = 0;
+  virtual bool canTranslate() const noexcept = 0;
+  virtual bool canEdit() const noexcept = 0;
```
Comment (TomTasche) [inline common/include/common/Document.h:23]: "editable" instead to get rid of can*?

File: odr/src/Document.cpp [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+    }
+    try {
+      return std::make_unique<ooxml::OfficeOpenXml>(storage);
+    } catch (...) {
+    }
+  } catch (access::NoZipFileException &) {
+  }
+  try {
+    FileMeta meta;
+    std::unique_ptr<access::ReadStorage> storage =
```
Comment (TomTasche) [inline odr/src/Document.cpp:37]: the code would be much easier to understand if this variable was called "encryptedStorage" or similar. took me a while to understand why almost the same code is used here as above

File: common/include/common/Document.h (hunk 1 above)
Comment (andiwand) [reply common/include/common/Document.h:23]: nice!

File: odr/src/Document.cpp (hunk 2 above)
Comment (andiwand) [reply odr/src/Document.cpp:37]: this storage could be holding an encrypted ooxml or legacy microsoft file. so `encryptedStorage` would be misleading.

**Pull request merged**